blockprod/src/detail/mod.rs

```diff
@@ -66,7 +69,141 @@ pub struct JobKey {
 }
 
 pub struct JobHandle {
-    cancel_signal: oneshot::Sender<()>,
+    cancel_sender: UnboundedSender<()>,
+}
+
+pub struct JobManager {
+    chainstate_handle: ChainstateHandle,
+    new_job_sender: UnboundedSender<(Id<GenBlock>, UnboundedSender<()>)>,
+    stop_job_sender: UnboundedSender<(Option<JobKey>, oneshot::Sender<bool>)>,
+    all_jobs: BTreeMap<JobKey, JobHandle>,
+}
+
+impl JobManager {
+    pub fn new(chainstate_handle: ChainstateHandle) -> JobManager {
+        let (new_job_sender, mut new_job_receiver) =
+            mpsc::unbounded_channel::<(Id<GenBlock>, UnboundedSender<()>)>();
+
+        let (stop_job_sender, mut stop_job_receiver) = mpsc::unbounded_channel();
+
+        let mut job_manager = JobManager {
+            chainstate_handle,
+            new_job_sender,
+            stop_job_sender,
+            all_jobs: BTreeMap::new(),
+        };
+
+        let (chainstate_sender, mut chainstate_receiver) =
+            mpsc::unbounded_channel::<Id<GenBlock>>();
+
+        job_manager.subscribe_to_chainstate(chainstate_sender);
+        job_manager.run(chainstate_receiver, new_job_receiver, stop_job_receiver);
+
+        job_manager
+    }
+
+    fn run(
+        &mut self,
+        mut chainstate_receiver: UnboundedReceiver<Id<GenBlock>>,
+        mut new_job_receiver: UnboundedReceiver<(Id<GenBlock>, UnboundedSender<()>)>,
+        mut stop_job_receiver: UnboundedReceiver<(Option<JobKey>, oneshot::Sender<bool>)>,
+    ) {
+        tokio::spawn(async move {
+            let mut all_jobs: BTreeMap<JobKey, JobHandle> = BTreeMap::new();
+
+            tokio::select! {
+                new_tip = chainstate_receiver.recv() => {
+                    let mut jobs_to_stop: Vec<JobKey> = vec![];
+
+                    for (job_key, _) in all_jobs.iter() {
+                        if job_key.current_tip != new_tip.unwrap() {
+                            jobs_to_stop.push(job_key.clone());
+                        }
+                    }
+
+                    for job_key in jobs_to_stop {
+                        if let Some(handle) = all_jobs.remove(&job_key) {
+                            handle.cancel_sender.send(());
+                        }
+                    }
+                }
+                new_job = new_job_receiver.recv() => {
+                    let new_job = new_job.unwrap();
+
+                    let job_key = JobKey {
+                        current_tip: new_job.0,
+                    };
+
+                    #[allow(clippy::map_entry)]
+                    if !all_jobs.contains_key(&job_key) {
+                        all_jobs.insert(
+                            job_key,
+                            JobHandle {
+                                cancel_sender: new_job.1,
+                            },
+                        );
+                    }
```

Comment:
You can return an error in a oneshot passed inside the new_job_receiver channel if the job is already there.